### PR TITLE
Add root/admin user detection

### PR DESCRIPTION
**Pull request by egg82**
This patch detects whether or not the server is currently executing as a privileged user and spits out a warning. The warning serves as a sort-of PSA for newer server admins who don't understand the risks of running as root.

We've seen plenty of bad/malicious plugins hit markets, and there's been a few close-calls with exploits in the past. Hopefully this helps mitigate some potential damage to servers, even if it is just a warning.

### CODE REVIEW

**Comment by egg82** [comment]
The method used is a little unconventional, so I'm open to feedback on that. I tested on a Windows 10 machine and Ubuntu 18.04 machine and it works on both.

File: Spigot-Server-Patches/0410-Add-root-admin-user-detection.patch [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
++import java.util.prefs.Preferences;
++
++public class ServerEnvironment {
++    private static final boolean runningAsRootOrAdmin;
++
++    static {
++        // https://stackoverflow.com/a/23538961
++        Preferences prefs = Preferences.systemRoot();
++        PrintStream err = System.err;
++        synchronized (err) { // Synchronized to prevent swapping back to "err" before printing the warning from the try/catch to console
```
Comment (Spottedleaf) [inline Spigot-Server-Patches/0410-Add-root-admin-user-detection.patch]: for the synchronized block to be effective, **if it even is effective at all**, you must re-check the current value of System.err while in it

Comment (egg82) [reply Spigot-Server-Patches/0410-Add-root-admin-user-detection.patch]: Did some research and testing, looks like a flush() will work fine. It's an odd situation, but..

**Comment by theminecoder** [comment]
I feel this should be bypassable in some cases (docker container, local dev server). Maybe we can skip it when using the same flag that skips the build date check.

**Comment by MiniDigger** [comment]
even in a docker container you should create a user, but I agree that a flag should be added. I highly disagree with reusing flags tho, that make them ambiguous.

**Comment by zachbr** [comment]
Not sure why you'd want to bypass this in any instance, if you're on some weird temporary setup you can just acknowledge and ignore the warning knowing it won't affect you.
👎 on any disable for this

**Comment by zachbr** [comment]
```
Patching vanilla jar...
Feb 19, 2020 2:14:18 PM java.util.prefs.FileSystemPreferences$2 run
WARNING: Could not create system preferences directory. System preferences are unusable.
System Info: Java 1.8 (OpenJDK 64-Bit Server VM 25.242-b08) Host:  Linux 3.10.0-1062.7.1.el7.x86_64 (amd64)
Loading libraries, please wait...
[14:14:31 INFO]: Starting minecraft server version 1.15.2
```

Followed by spamming this: 
```
[14:18:22 WARN]: Could not lock System prefs. Unix error code 32519.
[14:18:22 WARN]: Couldn't flush system prefs: java.util.prefs.BackingStoreException: Couldn't get file lock.
[14:18:52 WARN]: Could not lock System prefs. Unix error code 32519.
[14:18:52 WARN]: Couldn't flush system prefs: java.util.prefs.BackingStoreException: Couldn't get file lock.
[14:19:22 WARN]: Could not lock System prefs. Unix error code 32519.
[14:19:22 WARN]: Couldn't flush system prefs: java.util.prefs.BackingStoreException: Couldn't get file lock.
[14:19:52 WARN]: Could not lock System prefs. Unix error code 32519.
[14:19:52 WARN]: Couldn't flush system prefs: java.util.prefs.BackingStoreException: Couldn't get file lock
```

CentOS

**Comment by egg82** [comment]
https://stackoverflow.com/questions/23960451/java-system-preferences-under-different-users-in-linux
https://bugzilla.redhat.com/show_bug.cgi?id=741821

Hmm...

Is this on every CentOS box? Seems like a possible broken Java install.

**Pull request merged**